booster/library/Booster/Pattern/Binary.hs

```diff
@@ -71,9 +70,34 @@ data Version = Version
 instance Show Version where
     show version = printf "%d.%d.%d" version.major version.minor version.patch
 
+{- | The marshalling algorithm below unpacks terms into a global term
+   store, indexed by "shallow terms", i.e., where arguments in
+   application nodes are replaced by indexes into a global term store.
+
+@ShallowTerm@s can be either simple data without recursion
+(@DomainValue@, @Var@), or recursive data (@SymbolApplication@,
+@AndTerm@, @Injection@), where the arguments are initially @Int@s
+instead of @Term@s (internal collection types are constructed from the
+unpacked symbol applications later, and not expected to occur here).
+
+A lookup map for these @ShallowTerm@ indexes is maintained while
+unpacking blocks.
+
+Unpacked @ShallowTerm@s that are not found in the lookup map are added
+to both the lookup map (as @ShallowTerm@s) and to the term store (as
+@Term@s, resolving arguments), and the new index is returned.
+
+Conversely, when an unpacked @ShallowTerm@ has occurred before, the
+previously-added index is returned.
+-}
+newtype ShallowTerm = ShallowTerm (TermF Idx)
+    deriving (Eq, Ord, Show)
+
+type Idx = Int
+
 data Block
-    = BTerm Term
-    | BPredicate Predicate
+    = BTerm Idx
+    | BPredicate Idx -- Predicate -- problem?!?
```

Comment:
Is the problem here that we will use the same store for terms and predicates in `resolve -> mkSymbolApplication`? I think this is fine, as the terms and predicates indeed consist of the same bits and pieces ultimately.

Reply:
I should remove that comment, it was not a problem after all, we just wrap the `Bool` term into a `Predicate` at the end instead of during the procedure.